File: javalinfly-example/src/main/java/com/github/unldenis/UserHandler.java
package com.github.unldenis;


import com.github.unldenis.javalinfly.*;
import com.github.unldenis.javalinfly.annotation.Body;
import com.github.unldenis.javalinfly.annotation.Controller;
import com.github.unldenis.javalinfly.annotation.Get;
import com.github.unldenis.javalinfly.annotation.Path;
import com.github.unldenis.javalinfly.annotation.Post;
import com.github.unldenis.javalinfly.annotation.Query;
import io.javalin.http.Context;

import java.util.List;
import org.jetbrains.annotations.NotNull;

@Controller(path = "/user") public class UserHandler {

//    @Post(responseType = ResponseType.STRING)
//    public Response<String, String> createUser(Context ctx, @Body User user) {
//
//        return Response.err("this is my create error");
//    }

    @Post(tags = {"user"})
    public Response<Users, StandardError> createAll(Context ctx, @Body Users users) {
        return Response.ok(users);
    }



    @Get(tags = "user")
    public Response<User, StandardError> getUser(Context ctx, @Path String id, @Query @NotNull String age) {
        return Response.ok(new User("denis", 2));
    }

    public static class Users {
        public List<User> users;


        public Users() {}
        public Users(List<User> users) {
            this.users = users;
        }
    }

    public static class User {

        public @NotNull String name;
        public int age;


        public User(@NotNull String name, int age) {
            this.name = name;
            this.age = age;
        }
    }


    public static class StandardError {
        public @NotNull String cause;

        public StandardError(@NotNull String cause) {
            this.cause = cause;
        }
    }


}

File: javalinfly-example-kotlin/src/main/kotlin/UserHandler.kt
package com.quicklink.javalinfly.kotlin.example

import com.quicklink.javalinfly.ResponseType
import com.quicklink.javalinfly.annotation.*
import com.quicklink.javalinfly.kotlin.response
import io.javalin.http.Context

@Controller(path = "/user")
class UserHandler {

    @Post(tags = ["user"])
    fun createAll(ctx: Context, @Body users: Users) = response<Users, StandardError> {
        ok = Users()
    }

    @Get(responseType = ResponseType.STRING, tags = ["user"])
    fun getUser(ctx: Context, @Path id: String, @Query age: String?) = response<String, String> {

        if (age == null)
            err = "age is missing"

        ok = "id $id, age $age"

    }
    @Get(tags = ["user"])
    fun getAll(ctx: Context) = response<List<User>, StandardError> {
        ok = emptyList()

    }

    @Put
    fun putUser(ctx : Context, @Query userId : String, @Body user : User) = response<User, StandardError> {
        ok = user
    }
    class Users {
        var users: LinkedHashMap<String, User>? = null
    }

    class User {
        lateinit var name: String
        var age: Int = 0

        constructor()

        constructor(name: String, age: Int) {
            this.name = name
            this.age = age
        }
    }


    class StandardError(var cause: String, var details : String?)
}

